### ISSUE TITLE
Many extra processes

### ISSUE THREAD

**Issue by angrycandy** [tkf/emacs-jedi#330]
* With my old .python-environments at most two processes are created for each emacs:
```
/home/bfw/.emacs.d/.python-environments/i386/bin/python2 /home/bfw/.emacs.d/.python-environments/i386/bin/jediepcserver
/usr/bin/python2.7 /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/jedi/evaluate/compiled/subprocess/__main__.py /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages 2
```
```
(:emacs-version "25.3.1" :jedi-version "0.2.7")
((:version "2.7.15 (default, Oct  6 2018, 14:26:45) \n[GCC 5.4.0 20160603 (Red Hat 5.4.0-6)]" :name "sys" :file nil)
 (:version "0.12.1" :name "jedi" :file "/home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/jedi/__init__.pyc")
 (:version "0.0.5" :name "epc" :file "/home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/epc/__init__.pyc")
 (:version "0.0.3" :name "sexpdata" :file "/home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/sexpdata.pyc"))
```
* After removing the old env and doing a new install there are many processes. A new subprocess is created for every python file visited.
```
% mv ~/.emacs.d/.python-environments  ~/.emacs.d/.python-environments.hide
```
```
M-x jedi:install-server
```
```
/home/bfw/.emacs.d/.python-environments/i386/bin/python2 /home/bfw/.emacs.d/.python-environments/i386/bin/jediepcserver
/home/bfw/.emacs.d/.python-environments/i386/bin/python2 /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/jedi/evaluate/compiled/subprocess/__main__.py /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages 2.7.15
/home/bfw/.emacs.d/.python-environments/i386/bin/python2 /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/jedi/evaluate/compiled/subprocess/__main__.py /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages 2.7.15
/home/bfw/.emacs.d/.python-environments/i386/bin/python2 /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/jedi/evaluate/compiled/subprocess/__main__.py /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages 2.7.15
```
```
(:emacs-version "25.3.1" :jedi-version "0.2.7")
((:version "2.7.15 (default, Oct  6 2018, 14:26:45) \n[GCC 5.4.0 20160603 (Red Hat 5.4.0-6)]" :name "sys" :file nil)
 (:version "0.13.1" :name "jedi" :file "/home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/jedi/__init__.pyc")
 (:version "0.0.5" :name "epc" :file "/home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/epc/__init__.pyc")
 (:version "0.0.3" :name "sexpdata" :file "/home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/sexpdata.pyc"))
```
* *Is this expected behavior?*

**Comment by immerrr** [comment]
Nope, it's not expected.
What does `jedi:show-setup-info` say?

**Comment by angrycandy** [comment]
```;; Emacs Lisp version:
(:emacs-version "25.3.1" :jedi-version "0.2.7" :python-environment-version "0.0.2alpha0")
;; Python version:
((:version "2.7.15 (default, Oct  6 2018, 14:26:45) \n[GCC 5.4.0 20160603 (Red Hat 5.4.0-6)]" :name "sys" :file nil)
 (:version "0.13.1" :name "jedi" :file "/home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/jedi/__init__.pyc")
 (:version "0.0.5" :name "epc" :file "/home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/epc/__init__.pyc")
 (:version "0.0.3" :name "sexpdata" :file "/home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/sexpdata.pyc"))
;; Command line:
(:virtualenv "/usr/bin/virtualenv" :virtualenv-version "16.1.0\n")
;; Customization:
((jedi:complete-on-dot . t)
 (jedi:doc-display-buffer . display-buffer)
 (jedi:doc-hook view-mode)
 (jedi:doc-mode . rst-mode)
 (jedi:environment-root . "i386")
 (jedi:environment-virtualenv)
 (jedi:get-in-function-call-delay . 1000)
 (jedi:get-in-function-call-timeout . 3000)
 (jedi:goto-definition-config
  (nil nil nil)
  (t nil nil)
  (nil definition nil)
  (t definition nil)
  (nil nil t)
  (t nil t)
  (nil definition t)
  (t definition t))
 (jedi:goto-definition-marker-ring-length . 16)
 (jedi:imenu-create-index-function . jedi:create-nested-imenu-index)
 (jedi:import-python-el-settings . t)
 (jedi:install-imenu)
 (jedi:install-python-jedi-dev-command "pip" "install" "--upgrade" "git+https://github.com/davidhalter/jedi.git@master#egg=jedi")
 (jedi:key-complete .
                    [C-tab])
 (jedi:key-goto-definition .
                           [67108910])
 (jedi:key-goto-definition-pop-marker .
                                      [67108908])
 (jedi:key-related-names . "r")
 (jedi:key-show-doc . "d")
 (jedi:server-args)
 (jedi:server-command "/home/bfw/.emacs.d/.python-environments/i386/bin/jediepcserver")
 (jedi:setup-keys)
 (jedi:tooltip-method pos-tip popup)
 (jedi:use-shortcuts . t)
 (python-environment-default-root-name . "default")
 (python-environment-directory . "~/.emacs.d/.python-environments")
 (python-environment-virtualenv "virtualenv" "--system-site-packages" "--quiet"))
```

**Comment by immerrr** [comment]
Nothing stand out to me. What version does `M-x describe-package jedi-core` show?

**Comment by angrycandy** [comment]
```
jedi-core is a dependency package.

     Status: Installed in  jedi-core-20181112.123/  (unsigned).
    Version: 20181112.123
    Summary: Common code of jedi.el and company-jedi.el
   Requires: emacs-24, epc-0.1.0, python-environment-0.0.2, cl-lib-0.5
Required by: jedi-20160426.456
```

**Comment by immerrr** [comment]
Wait, so you are saying there are multiple jedi subprocesses, not various jediepcserver processes?

I think I have fixed this with `20181117` version. Could you try that instead? I saw that behaviour once, introduced a fix, but then I was unable to reproduce it reliably to be sure that the fix actually works. 

Do you have a reliable repro for the spawning subprocesses situation? Could you share it?

**Comment by angrycandy** [comment]
I have a very reliable repro. It's on a work computer, so I can't share it. I'll try `20181117` tomorrow and let you know how it goes.

**Comment by angrycandy** [comment]
Using jedi-core-20181117.954 did not fix the problem.

I removed the old installation.
```
% ls .emacs.d/elpa
archives		    concurrent-20161229.330  deferred-20170901.1330  jedi-20160426.456	     popup-20160709.1429
auto-complete-20170125.245  ctable-20171006.11	     epc-20140610.534	     jedi-core-20181112.123  python-environment-20150310.853
% rm -rf .emacs.d/elpa .emacs.d/.python-environments
% cat .emacs.el
(setq package-archives
      '(("gnu" . "https://elpa.gnu.org/packages/")
        ("melpa" . "https://melpa.org/packages/")))
(setq jedi:environment-root
      (replace-regexp-in-string
       "\n" ""
       (shell-command-to-string "uname -i")))
(setq jedi:complete-on-dot t)
(setq jedi:use-shortcuts t) ;; M-. & M-, for goto def & pop back
(defun my-python-mode-hook ()
  (jedi:setup)
  (local-set-key (kbd "C-c s") 'jedi:stop-server))
(add-hook 'python-mode-hook 'my-python-mode-hook)
% 
```
Ran emacs to package install jedi and jedi:install-server:
```
% ls .emacs.d/elpa
archives		    concurrent-20161229.330  deferred-20170901.1330  jedi-20160426.456	     popup-20160709.1429
auto-complete-20170125.245  ctable-20171006.11	     epc-20140610.534	     jedi-core-20181117.954  python-environment-20150310.853
% 
```
After visiting three python files:
```
% ps -wwfu bfw
UID        PID  PPID  C STIME TTY          TIME CMD
bfw       4541 29145  0 05:35 pts/0    00:00:00 emacs
bfw      16763  4541  1 05:37 pts/1    00:00:00 /home/bfw/.emacs.d/.python-environments/i386/bin/python2 /home/bfw/.emacs.d/.python-environments/i386/bin/jediepcserver
bfw      16767 16763  0 05:37 pts/1    00:00:00 /home/bfw/.emacs.d/.python-environments/i386/bin/python2 /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/jedi/evaluate/compiled/subprocess/__main__.py /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages 2.7.15
bfw      18689 16763  0 05:37 pts/1    00:00:00 /home/bfw/.emacs.d/.python-environments/i386/bin/python2 /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/jedi/evaluate/compiled/subprocess/__main__.py /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages 2.7.15
bfw      18737 16763  0 05:37 pts/1    00:00:00 /home/bfw/.emacs.d/.python-environments/i386/bin/python2 /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/jedi/evaluate/compiled/subprocess/__main__.py /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages 2.7.15
bfw      18772 16763  0 05:37 pts/1    00:00:00 /home/bfw/.emacs.d/.python-environments/i386/bin/python2 /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/jedi/evaluate/compiled/subprocess/__main__.py /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages 2.7.15
bfw      18843 16763  0 05:37 pts/1    00:00:00 /home/bfw/.emacs.d/.python-environments/i386/bin/python2 /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/jedi/evaluate/compiled/subprocess/__main__.py /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages 2.7.15
bfw      18955 16763  1 05:37 pts/1    00:00:00 /home/bfw/.emacs.d/.python-environments/i386/bin/python2 /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/jedi/evaluate/compiled/subprocess/__main__.py /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages 2.7.15
bfw      19663 16763  1 05:37 pts/1    00:00:00 /home/bfw/.emacs.d/.python-environments/i386/bin/python2 /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/jedi/evaluate/compiled/subprocess/__main__.py /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages 2.7.15
```
```
jedi-core is a dependency package.

     Status: Installed in  jedi-core-20181117.954/  (unsigned).
    Version: 20181117.954
    Summary: Common code of jedi.el and company-jedi.el
   Requires: emacs-24, epc-0.1.0, python-environment-0.0.2, cl-lib-0.5
Required by: jedi-20160426.456
```

**Comment by angrycandy** [comment]
With the new packages installed I replaced the new env with an old, working env.
```
% mv .emacs.d/.python-environments .emacs.d/.python-environments.new
% ln -s ~/.emacs.d/.python-environments.hide .emacs.d/.python-environments
%
```
And I'm back to just two procs per emacs no matter how many python files I visit:
```
bfw      26969 23510  1 07:27 pts/1    00:00:00 /home/bfw/.emacs.d/.python-environments/i386/bin/python2 /home/bfw/.emacs.d/.python-environments/i386/bin/jediepcserver
bfw      27731 26969  0 07:27 pts/1    00:00:00 /usr/bin/python2.7 /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages/jedi/evaluate/compiled/subprocess/__main__.py /home/bfw/.emacs.d/.python-environments/i386/lib/python2.7/site-packages 2

```
I think the difference is what `jedi:install-server` does now compare to what it did on Sep 30 when I created the env that works better.

**Comment by immerrr** [comment]
It's not `jedi:install-server` per se, it's just the new version of jediepcserver.py that works like that. I'll have another look.

**Comment by angrycandy** [comment]
@immerrr are you able to reproduce this problem?

I reproduced it in cloud shell with my google account.

https://console.cloud.google.com/cloudshell

**Comment by immerrr** [comment]
I'll try to look at tomorrow. If you have a repro from a clean environment, it would help a lot.

**Comment by angrycandy** [comment]
I sent you email with the password for a google account I created that repros the problem in cloudshell. Let me know here if you have problems as I don't plan to monitor the burner account.

**Comment by angrycandy** [comment]
Maybe another clue: jedi:stop-server doesn't kill all jedi.el subprocesses after visiting some python files. This is reproducible in the cloudshell account mentioned above.

**Comment by immerrr** [comment]
Hi, thank you for your patience. I have managed to reproduce it, you can shut down the cloud shell, thank you.

**Comment by asmeurer** [comment]
I am seeing this as well. Emacs sometimes starts up Python processes until the system process limit is filled. Killing emacs fixes it. I can run some debugging thing next time it happens if it would help.

**Comment by immerrr** [comment]
Ok, it's a bug in Jedi in the end.

Here's the most detailed report on it: https://github.com/davidhalter/jedi/issues/1242
Here's a PR that slowly pushes things forward: https://github.com/davidhalter/jedi/pull/1238

I'll see what I can do to mitigate it in jediepcserver.

**Comment by immerrr** [comment]
Or that, yes

**Comment by immerrr** [comment]
Ok, try new master (it might take a few hours for melpa to catch up).

**Comment by angrycandy** [comment]
I updated to jedi-core-20181207.1 and it fixed the problem for me. Thanks the quick response and resolution!

**Comment by immerrr** [comment]
We can close the issue then, awesome!
